File: govbr/src/main/java/br/serpro/cidadao/govbr/GovBrWebView.java
package br.serpro.cidadao.govbr;

import android.content.Context;
import android.util.AttributeSet;
import android.webkit.*;

import java.math.BigInteger;
import java.security.SecureRandom;

public class GovBrWebView extends WebView {

    Context context;

    public static final String DESENV = "http://sso.staging.acesso.gov.br";
    public static final String HOMOLOGACAO = "http://sso.validacao.acesso.gov.br";
    public static final String PRODUCAO = "http://sso.acesso.gov.br";

    public GovBrWebView(Context context) {
        super(context);
        initView(context);
    }

    public GovBrWebView(Context context, AttributeSet attrs) {
        super(context, attrs);
        initView(context);
    }

    /**
     * @param ambiente
     */
    public void loginNoGovBr(String ambiente, GovBrWebViewClient client) {
        setWebViewClient(client);

        loadUrl(this.getUrlProvider(ambiente));
    }

    /**
     * @param context
     */
    private void initView(Context context) {
        this.context = context;
        getSettings().setAppCacheEnabled(false);
        getSettings().setCacheMode(WebSettings.LOAD_NO_CACHE);
        getSettings().setSaveFormData(false);
        clearCache(true);
        clearHistory();
        clearFormData();
        getSettings().setJavaScriptEnabled(true);

        setWebChromeClient(new WebChromeClient() {
            @Override
            public void onProgressChanged(WebView view, int progress) {
                //setProgress(progress * 100);
            }
        });
    }

    private String getUrlProvider(String urlProvider) {
        return urlProvider +
                "/authorize?response_type=code&client_id=" + context.getString(R.string.gov_br_client_id) +
                "&scope=" + context.getString(R.string.gov_br_scopes) +
                "&redirect_uri=" + context.getString(R.string.gov_br_redirect_url) +
                "&nonce=" + createRandomNumber() +
                "&state=" + createRandomNumber();
    }

    private static String createRandomNumber() {
        return new BigInteger(50, new SecureRandom()).toString(16);
    }

    public void onDestroy() {
        CookieSyncManager.createInstance(context);
        CookieManager cookieManager = CookieManager.getInstance();
        cookieManager.removeAllCookie();
    }

}

File: govbr/src/main/java/br/serpro/cidadao/govbr/GovBrWebView.kt
package br.serpro.cidadao.govbr

import android.annotation.SuppressLint
import android.content.Context
import android.util.AttributeSet
import android.webkit.*

class GovBrWebView : WebView {

    constructor(context: Context) : super(context) {
        initView()
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        initView()
    }

    /**
     * @param ambiente
     */
    fun loginNoGovBr(ambiente: String, client: GovBrWebViewClient) {
        webViewClient = client

        loadUrl(this.getUrlProvider(ambiente))
    }

    @SuppressLint("SetJavaScriptEnabled")
    private fun initView() {
        clearWebView()

        settings.setAppCacheEnabled(false)
        settings.cacheMode = WebSettings.LOAD_NO_CACHE
        settings.saveFormData = false
        settings.javaScriptEnabled = true

        //Corrige o problema do Android Webview: “Uncaught TypeError: Cannot read property 'getItem' of null”
        settings.domStorageEnabled = true
    }

    private fun getUrlProvider(urlProvider: String): String {
        return urlProvider +
                "/authorize?response_type=code&client_id=" + context?.getString(R.string.gov_br_client_id) +
                "&scope=" + context?.getString(R.string.gov_br_scopes) +
                "&redirect_uri=" + context?.getString(R.string.gov_br_redirect_url) +
                "&nonce=" + randomString(12) +
                "&state=" + randomString(13)
    }

    private fun clearCookies() {
        CookieSyncManager.createInstance(context)
        val cookieManager = CookieManager.getInstance()
        cookieManager.removeAllCookie()
    }

    fun clearWebView() {
        clearCache(true)
        clearHistory()
        clearFormData()
        clearView()

        clearCookies()
    }

    companion object {
        const val DESENV = "http://sso.staging.acesso.gov.br"

        //Usando ambiente de staging para homologação enquanto não estabiliza o desenvolvimento do Gov.br
        const val HOMOLOGACAO = "http://sso.staging.acesso.gov.br"
        //const val HOMOLOGACAO = "http://sso.validacao.acesso.gov.br"

        const val PRODUCAO = "http://sso.acesso.gov.br"

        private fun randomString(length: Int): String {
            val letters = "abcdefghijklmnopqrstuvwxyz0123456789"
            return (1..length)
                .map { kotlin.random.Random.nextInt(0, letters.length) }
                .map(letters::get)
                .joinToString("")
        }
    }

}
